packages/web/src/lib/core/auth/internal/StitchAuthImpl.ts

```diff
@@ -56,6 +85,129 @@ export default class StitchAuthImpl extends CoreStitchAuth<StitchUser>
     );
   }
 
+  public loginWithRedirect(credential: StitchRedirectCredential) {
+    const { redirectUrl, state } = this.prepareRedirect(
+      credential,
+      RedirectTypes.Login
+    );
+
+    window.location.replace(
+      this.browserAuthRoutes.getAuthProviderRedirectRoute(
+        credential,
+        redirectUrl,
+        state,
+        this.deviceInfo
+      )
+    );
+  }
+
+  public linkUserWithRedirect(
+    user: StitchUser,
+    credential: StitchRedirectCredential
+  ) {
+    this.storage.set(
+      RedirectKeys.LinkUser,
+      JSON.stringify(new StitchUserCodec(this).encode(user))
+    );
+
+    const { redirectUrl, state } = this.prepareRedirect(
+      credential,
+      RedirectTypes.Login
+    );
+
+    const link = this.browserAuthRoutes.getAuthProviderLinkRedirectRoute(
+      credential,
+      redirectUrl,
+      state,
+      this.deviceInfo
+    );
+
+    fetch(
+      new Request(link, {
+        credentials: "include",
+        headers: {
+          Authorization: "Bearer " + this.authInfo.accessToken
+        }
+      })
+    ).then(response => {
+      window.location.replace(response.headers.get("X-Stitch-Location")!);
+    });
+  }
+
+  public hasRedirect(): boolean {
+    return !!this.storage.get(RedirectKeys.State);
+  }
+
+  public handleRedirect(): Promise<StitchUser> {
+    const cleanup = () => {
+      this.storage.remove(RedirectKeys.State);
+      this.storage.remove(RedirectKeys.RedirectProvider);
+    };
+
+    if (typeof window === "undefined") {
+      // This means we're running in some environment other
+      // than a browser - so handling a redirect makes no sense here.
+      cleanup();
+      return Promise.reject();
+    }
+
+    if (!window.location || !window.location.hash) {
+      cleanup();
+      return Promise.reject();
+    }
+
+    const ourState = this.storage.get(RedirectKeys.State);
+    const redirectProvider = this.storage.get(RedirectKeys.RedirectProvider);
+
+    cleanup();
+
+    const redirectFragment = window.location.hash.substring(1);
+    const redirectState = this.parseRedirectFragment(
+      redirectFragment,
+      ourState
+    );
+
+    if (redirectState.lastError || (redirectState.found && !redirectProvider)) {
+      // this._error = redirectState.lastError;
+      window.history.replaceState(null, "", this.pageRootUrl());
```

Comment:
Comment why we do this (to remove the fragment). Ensure that this actually removes the fragment. It may not.